Create CurrentUserTests.cs

Pasha1968 opened the pull request:
Created CurrentUserTests.cs - tests for CurrentUserService.cs

Pasha1968 committed:
Create CurrentUserTests.cs

gingerbread-monster commented on CharlieBackend.Api.UnitTes/CurrentUserTests.cs:
```diff
@@ -0,0 +1,97 @@
+        [Fact]
+        public void getAccountId()
+        {
+            var _httpContextAccessorMock = new Mock<IHttpContextAccessor>();
+            var httpContext = new DefaultHttpContext();
+            var mockIdentity = new Mock<GenericIdentity>("Vasya");
```
I think you can move this repetitive boilerplate code from each test method into `CurrentUserTests` class constructor, can't you?

gingerbread-monster replied:
Partially solved in https://github.com/ita-social-projects/WhatBackend/pull/504/commits/e24ea4bfff6c941a77091f4f006367542332ab6b

gingerbread-monster commented on CharlieBackend.Api.UnitTes/CurrentUserTests.cs:
```diff
@@ -0,0 +1,97 @@
+            var currentuserService = new CurrentUserService(_httpContextAccessorMock.Object);
+            var Email = currentuserService.Email;
+
+            Assert.Equal("Hello@mail.com", Email);
```
Also I think we're supposed to use **FluentAssertions** for tests. You can view `MentorServiceTests` for reference.

e.g.:
https://github.com/ita-social-projects/WhatBackend/blob/899f397be4e51f81fc156f059c3ced0a0a60d669/CharlieBackend.Api.UnitTes/MentorServiceTests.cs#L65

gingerbread-monster replied:
Solved in https://github.com/ita-social-projects/WhatBackend/pull/504/commits/e24ea4bfff6c941a77091f4f006367542332ab6b

Pasha1968 committed:
add constructor

gingerbread-monster commented on CharlieBackend.Api.UnitTes/CurrentUserTests.cs:
```diff
@@ -0,0 +1,97 @@
+                .Returns(new Claim[] {
+                    new Claim("Email", "Hello@mail.com") 
+                });
+            _httpContext.User = new ClaimsPrincipal(_mockIdentity.Object);
+            _httpContextAccessorMock.Setup(req => req.HttpContext).Returns(_httpContext);
```
Aren't these 2 lines can be moved from each method into constructor as well? :)

gingerbread-monster replied:
Solved in https://github.com/ita-social-projects/WhatBackend/pull/504/commits/599f172a458febfc3f38a1fd66639d04ac39d823

gingerbread-monster commented on CharlieBackend.Api.UnitTes/CurrentUserTests.cs:
```diff
@@ -0,0 +1,97 @@
+            _httpContextAccessorMock.Setup(req => req.HttpContext)
+                .Returns(new DefaultHttpContext());
+
+            Func<long> getId = () => _currentuserService.AccountId;
+            Action act = () => getId();
```
Why add redundant delegates? Wouldn't it already work with the first one?

```suggestion
            Func<long> act = () => _currentuserService.AccountId;
```

gingerbread-monster replied:
Solved in https://github.com/ita-social-projects/WhatBackend/pull/504/commits/599f172a458febfc3f38a1fd66639d04ac39d823

gingerbread-monster commented on CharlieBackend.Api.UnitTes/CurrentUserTests.cs:
```diff
@@ -0,0 +1,97 @@
+        [Fact]
+        public void getAccountId()
+        {
+            _mockIdentity.Setup(i => i.Claims)
+                .Returns(new Claim[]{
+                    new Claim("AccountId", "123") 
+                });
+            _httpContext.User = new ClaimsPrincipal(_mockIdentity.Object);
+            _httpContextAccessorMock.Setup(req => req.HttpContext).Returns(_httpContext);
+
+            _currentuserService.AccountId.Should().Be(123);
```
When introducing expected values as constants, what I would suggest is declaring them in a single place (at least in the `Arrange` section of the test). This way you're decreasing the possibility of making a mistake and making code more "fluent".
Something like this:

```cs
            // Arrange
            long expectedAccountId = 123;
            
            _mockIdentity.Setup(i => i.Claims)
                .Returns(new Claim[]{
                    new Claim("AccountId", expectedAccountId.ToString()) 
                });

            // Act & Assert
            _currentuserService.AccountId.Should().Be(expectedAccountId);
```

This comment also applies to other methods & constants of yours.

gingerbread-monster replied:
Solved in https://github.com/ita-social-projects/WhatBackend/pull/504/commits/599f172a458febfc3f38a1fd66639d04ac39d823

Pasha1968 committed:
Update CurrentUserTests.cs

deleted one delegate, moved 2 lines to constructor, moved constant's declaration into Arrange section

andrii-developer commented on CharlieBackend.Api.UnitTes/CurrentUserTests.cs:
```diff
@@ -0,0 +1,100 @@
+        }
+
+        [Fact]
+        public void ThrowErrorIfNotLoggined()
```
Please try to follow [best practices](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices?ranMID=46131&ranEAID=a1LgFw09t88&ranSiteID=a1LgFw09t88-PBLgRQP1_o9AbEiA2hJ8zg&epi=a1LgFw09t88-PBLgRQP1_o9AbEiA2hJ8zg&irgwc=1&OCID=AID2000142_aff_7806_1243925&tduid=%28ir__nnr3tb2a2okfqj1hkk0sohz3132xusa63bxexe1w00%29%287806%29%281243925%29%28a1LgFw09t88-PBLgRQP1_o9AbEiA2hJ8zg%29%28%29&irclickid=_nnr3tb2a2okfqj1hkk0sohz3132xusa63bxexe1w00#naming-your-tests) for naming of your tests. 
Please rename test methods like this one:
```suggestion
        public void AccountId_UserNotLoggedIn_ThrowsUnauthorizedAccessException()
```

andrii-developer commented on CharlieBackend.Api.UnitTes/CurrentUserTests.cs:
```diff
@@ -0,0 +1,100 @@
+        }
+
+        [Fact]
+        public void getAccountId()
```
Please follow [C# coding conventions](https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions). 
When naming public members of types, such as fields, properties, events, methods, and local functions, use pascal casing.

andrii-developer commented on CharlieBackend.Api.UnitTes/CurrentUserTests.cs:
```diff
@@ -0,0 +1,100 @@
+
+            // Act & Assert
+            _currentuserService.Email.Should().BeEquivalentTo(expectedEmail);
+        }
```
Please add one blank line between methods.

andrii-developer reviewed (changes requested):
Waiting for the author to address comments.

gingerbread-monster commented on CharlieBackend.Api.UnitTes/CurrentUserTests.cs:
```diff
@@ -42,56 +44,57 @@ public void ThrowErrorIfNotLoggined()
+            _currentuserService.AccountId.Should().Be(expectedAccountId);
         }
         [Fact]
         public void getEntityId()
```
You're the first person I see to completely ignore the coding style of the rest of the project  and use _camelCase_ methods naming. 
Come on, it's not JavaScript. We use _PascalCase_ naming in C#.

Even better if you'd follow [the unit test naming convention](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#naming-your-tests). You can also look at other unit tests for reference.

gingerbread-monster replied:
Duplicate of https://github.com/ita-social-projects/WhatBackend/pull/504#discussion_r646455509

Pasha1968 committed:
Update CurrentUserTests.cs

Added one blank line between methods;
Renamed method using PascalCase

andrii-developer commented on CharlieBackend.Api.UnitTes/CurrentUserTests.cs:
```diff
@@ -0,0 +1,108 @@
+        }
+
+        [Fact]
+
```
Remove empty lines between [Fact] attribute and methods.

Pasha1968 committed:
Update CurrentUserTests.cs

removed empty lines between [Fact] and methods

andrii-developer commented on CharlieBackend.Api.UnitTes/CurrentUserTests.cs:
```diff
@@ -0,0 +1,103 @@
+using CharlieBackend.Core.Entities;
+using CharlieBackend.Business.Services;
+using FluentAssertions;
+
```
Remove this empty line.

Pasha1968 committed:
Update CurrentUserTests.cs

removed empty line
